### PR TITLE
Fixed SI-6353: applyDynamic with sugared applications

### PR DESCRIPTION
**Pull request by namin**
- Accept sugared applications such as x(1) if x implements Dynamic,
  so x(1) gets re-written to x.apply(1).
- When picking a dynamic rewrite for x.apply(1), favor applyDynamic
  instead of the default selectDynamic.

Review by @paulp or @gkossakowski.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -3842,11 +3843,17 @@ trait Typers extends Modes with Adaptations with Tags {
           }
           @inline def hasNamedArg(as: List[Tree]) = as.collectFirst{case AssignOrNamedArg(lhs, rhs) =>}.nonEmpty
 
+          def desugaredApply = {
+            tree match {
+              case Select(`qual`, nme.apply) => true
+              case _ => false
+            }
+          }
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: `TreeInfo` would probably be a better place for this method

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -1107,10 +1107,11 @@ trait Typers extends Modes with Adaptations with Tags {
         case _ =>
           def applyPossible = {
             def applyMeth = member(adaptToName(tree, nme.apply), nme.apply)
+            dyna.acceptsApplyDynamic(tree.tpe) || (
             if ((mode & TAPPmode) != 0)
               tree.tpe.typeParams.isEmpty && applyMeth.filter(!_.tpe.typeParams.isEmpty) != NoSymbol
             else
-              applyMeth.filter(_.tpe.paramSectionCount > 0) != NoSymbol
+              applyMeth.filter(_.tpe.paramSectionCount > 0) != NoSymbol)
           }
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: To be honest, I don't feel like the absense of indentation is justified here.

Comment (namin) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Done.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -1107,10 +1107,11 @@ trait Typers extends Modes with Adaptations with Tags {
         case _ =>
           def applyPossible = {
             def applyMeth = member(adaptToName(tree, nme.apply), nme.apply)
+            dyna.acceptsApplyDynamic(tree.tpe) || (
             if ((mode & TAPPmode) != 0)
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I think indentation should be changed to make this change and code clearer.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (namin) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Isn't it too specific?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 4]
```diff
@@ -3842,11 +3843,17 @@ trait Typers extends Modes with Adaptations with Tags {
           }
           @inline def hasNamedArg(as: List[Tree]) = as.collectFirst{case AssignOrNamedArg(lhs, rhs) =>}.nonEmpty
 
+          def desugaredApply = {
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: No need for braces.
The `def desugaredApply = tree match {` will do.

Comment (namin) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Done.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I think "a selection of apply from something" is more or less generic. There are similar thingies in TreeInfo.

Comment (namin) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Yes, but here I check that the something is what I expect... so it's a bit more specific. I prefer to leave it here, unless you or others insist.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (namin) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Done.

**Comment by gkossakowski** [comment]
Great work and great commit message. LGTM. Thanks!

**Pull request merged**